### PR TITLE
Pro 2490 cache on demand

### PR DESCRIPTION
**Pull request by ValJed**
[PRO-2490](https://linear.app/apostrophecms/issue/PRO-2490/integrate-the-express-cache-on-demand-middleware-in-a3-core)

## Summary

Adds Express cache on deman on get routes for pieces and pages (getOne / getAll). Also on the main route for rendering pages.

Has been tested with refactored `cache-on-demand` and `express-cache-on-demand`. See these PRs:
* [cache-on-demand](https://github.com/apostrophecms/cache-on-demand/pull/2)
* [express-cache-on-demand](https://github.com/apostrophecms/express-cache-on-demand/pull/5)

⚠️ This one should wait that both above have been merged and released.

Tested with 50 concurrent users and 200 reps, on benchmark mode (so without delays between requests), we can see an obvious difference on the Elapsed time, an average of 20 seconds (even more on the example) of difference between requests with cache on demand and without.
image

image

## What are the specific steps to test this change?

> 1. Can be tested with the PRs above on testbed
> 2. Run the project and run siege on pages.

## What kind of change does this PR introduce?

- [ ] Bug fix
- [X] New feature
- [ ] Refactor
- [ ] Documentation
- [ ] Build-related changes
- [ ] Other

## Make sure the PR fulfills these requirements:

- [X] It includes a) the existing issue ID being resolved, b) a convincing reason for adding this feature, or c) a clear description of the bug it resolves
- [X] The changelog is updated
- [ ] Related documentation has been updated (Will be done in A3-docs)
- [ ] Related tests have been updated

If adding a new feature without an already open issue, it's best to open a **feature request issue** first and wait for approval before working on it.

**Other information:**

### CODE REVIEW

File: modules/@apostrophecms/page/index.js [hunk 1]
```diff
@@ -77,95 +82,102 @@ module.exports = {
...
 
-      async getAll(req) {
-        self.publicApiCheck(req);
-        const all = self.apos.launder.boolean(req.query.all);
-        const archived = self.apos.launder.booleanOrNull(req.query.archived);
-        const flat = self.apos.launder.boolean(req.query.flat);
-        const autocomplete = self.apos.launder.string(req.query.autocomplete);
+      getAll: [
+        ...self.enableCacheOnDemand ? [ expressCacheOnDemand ] : [],
+        async (req) => {
```
Comment (ValJed) [inline modules/@apostrophecms/page/index.js:87]: Thins func doesn't change just indentation.

File: modules/@apostrophecms/piece-type/index.js [hunk 2]
```diff
@@ -164,79 +165,91 @@ module.exports = {
...
-        delete newInstance._url;
-        return newInstance;
+  restApiRoutes(self) {
+    const { enableCacheOnDemand = true } = self.apos
+      .modules['@apostrophecms/express'].options;
+
+    return {
+      getAll: [
+        ...enableCacheOnDemand ? [ expressCacheOnDemand ] : [],
+        async (req) => {
```
Comment (ValJed) [inline modules/@apostrophecms/piece-type/index.js:175]: Thins func doesn't change just indentation.

**Review by boutell — COMMENTED** [review]
This looks good but not officially approving until that bug you found in express-cache-on-demand gets fixed (see my review).

**Review by boutell — APPROVED** [review]

File: package.json [hunk 3]
```diff
@@ -61,6 +61,7 @@
     "eslint-plugin-promise": "^5.1.0",
     "express": "^4.16.4",
     "express-bearer-token": "^2.4.0",
+    "express-cache-on-demand": "^1.0.2",
```
Comment (boutell) [inline package.json:64]: On release day let's make sure we bump this to the new, debugged release of `express-cache-on-demand` as a minimum just to be thorough.

**Review by boutell — APPROVED** [review]

**Pull request merged**